Default settings for the multi block

james-pearce-bath-ac opened the pull request:
This is to add some default settings to the multi block, including a block title, default presentation style and default set of blocks to add to the multi block on creation.

ma2716 committed:
Adding the new development features (multiblock name and presentation style settings).

marxjohnson commented on block_multiblock.php:
```diff
@@ -62,6 +75,8 @@ public function applicable_formats() {
     public function specialization() {
         if (isset($this->config->title)) {
             $this->title = format_string($this->config->title, true, ['context' => $this->context]);
+        } else if (isset(get_config('block_multiblock')->title)) {
```
Rather than doing 2 calls to get_config() here, it would be neater to do something like:
```php
$defaulttitle = get_config('block_multiblock', 'title');
if (isset($this->config_title)) {
   ...
} else if ($defaulttitle) {
   $this->title = format_string($defaulttitle, true, ['context' => $this->context]);
}
```
(get_config returns false if the requested config value is not set)

james-pearce-bath-ac replied:
I've updated the way we handle the default title, taking it out of the settings screens and intending to use the language customisation. This code is now not used.

marxjohnson commented on block_multiblock.php:
```diff
@@ -92,31 +106,61 @@ public function load_multiblocks($contextid) {
+            return $this->content;
+        }
+
+        $context = $DB->get_record('context', ['contextlevel' => CONTEXT_BLOCK, 'instanceid' => $this->instance->id]);
```
You should use `context_block::get_instance($this->instance->id)` to get the context, rather than doing a database query directly.

james-pearce-bath-ac replied:
Changing the code to your suggestion failed so have reverted to the direct db request.

marxjohnson commented on block_multiblock.php:
```diff
@@ -134,15 +178,27 @@ public function get_content() {
+        }
+        if (!empty($this->config->presentation)) {
+            $template = $this->config->presentation;
+        } else if (isset(get_config('block_multiblock')->presentation)) {
```
Again, it would be neater to do a single `get_config()` call

james-pearce-bath-ac replied:
I have updated this code to make it a single request.

marxjohnson commented on block_multiblock.php:
```diff
@@ -280,8 +335,14 @@ public static function get_valid_presentations(): array {
+        foreach ($presentations as $presentationid => $presentation) {
+            array_push($multiblockpresentationoptions, $presentationid);
+        }
+        if (isset(get_config('block_multiblock')->presentation)) {
```
As before, a single `get_config()` would be neater.

james-pearce-bath-ac replied:
I have updated this code to make it a single request.

marxjohnson commented on block_multiblock.php:
```diff
@@ -280,8 +335,14 @@ public static function get_valid_presentations(): array {
+        }
+        if (isset(get_config('block_multiblock')->presentation)) {
+            return $multiblockpresentationoptions[get_config('block_multiblock')->presentation];
+        } else if (isset($presentations['accordion'])) {
```
If I understand correctly, this now changes the default presentation from "tabbed-list" to "accordion" if none is set. Was that intentional? If this is just Bath's preference, I think it would be better to retain the current default so we dont unexpectedly change behaviour for other users, especially as we are adding an option to set a different default.

james-pearce-bath-ac replied:
I have changed the code so the default is not changed for the presentation style.

marxjohnson commented on classes/auto/adddefaultblock.php:
```diff
@@ -0,0 +1,98 @@
+ * Form for adding a block to a multiblock instance.
+ *
+ * @package   block_multiblock
+ * @copyright 2019 Peter Spicer <peter.spicer@catalyst-eu.net>
```
As this is a new file, it doesn't need Peter's name on the copyright. I'm not sure exactly how it works for Bath, but I would guess that @copyright should be University of Bath, and @author should be you.

james-pearce-bath-ac replied:
updated

marxjohnson commented on classes/auto/adddefaultblock.php:
```diff
@@ -0,0 +1,98 @@
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace block_multiblock\auto;
```
According to moodle's namespace rules, all second-level namespaces are reserved for core APIs. If you want a custom namespace in your plugin, it needs to be under "local" (so in this case, `block_multiblock\local\auto` in `blocks/multiblock/classes/local/auto/`). However, in this case it might be simpler just to use the top-level `block_multiblock` namespace and put it in `blocks/multiblock/classes`.

marxjohnson commented on lang/en/block_multiblock.php:
```diff
@@ -41,6 +42,12 @@
 
+$string['moodle_multiblock_title'] = 'Multiblock Title';
+$string['moodle_multiblock_title_desc'] = 'This title will be displayed as the heading of multiblock in the course.';
+$string['moodle_multiblock_presentation_style'] = 'Multiblock presentation style';
```
I don't see any particular reason to prefix these string IDs with "moodle_".

james-pearce-bath-ac replied:
Language strings have been updated.

marxjohnson commented on classes/auto/adddefaultblock.php:
```diff
@@ -0,0 +1,98 @@
+/**
+ * Libeary for adding a block to a multiblock instance.
+ *
+ * @package   block_multiblock
```
You dont need separate license and copyright info in the class's docblock

james-pearce-bath-ac replied:
updated

marxjohnson commented on edit_form.php:
```diff
@@ -21,7 +21,7 @@
  * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
-
+use block_multiblock\helper;
```
This doesn't appear to actually be used in this file, so can be removed.

james-pearce-bath-ac replied:
this has been removed

marxjohnson commented on edit_form.php:
```diff
@@ -40,13 +40,13 @@ class block_multiblock_edit_form extends block_edit_form {
         // First, the block's title.
         $mform->addElement('text', 'config_title', get_string('blocktitle', 'block_multiblock'));
         $mform->setType('config_title', PARAM_TEXT);
+        $mform->setDefault('config_title', get_config('block_multiblock')->title);
```
As you're only getting a single value, `get_config('block_multiblock', 'title')` would be neater.

marxjohnson commented on settings.php:
```diff
@@ -0,0 +1,62 @@
+
+    // Multiblock title (heading).
+    $settings->add(new admin_setting_configtext('block_multiblock/title', get_string('moodle_multiblock_title', 'block_multiblock'),
+    get_string('moodle_multiblock_title_desc', 'block_multiblock'), get_string('pluginname_recommended', 'block_multiblock'),
```
I don't think its appropriate to set a default title like this, as its specific to Bath's use case. Better to leave it blank so it defaults to "Multiblock" unless it's set. This also means removing the lang string and reverting the behat tests.

james-pearce-bath-ac replied:
I have removed this code for handling the block title so the language strings are the way to now define the block title.

marxjohnson replied:
Hi James,
Sorry it might not have been clear what I meant, I was specifically talking about the use of the 'pluginname_reccommended' string to set a default title if none is set. The name "Course Toolkit" is specific to Bath's use case so it doesn't make sense for this to be the fallback for everyone. Setting this via a lang string instead works, but I'm happy to keep the setting here for the additional flexibility it provides, just have it default to an empty string, and have the block use "Multiblock" as the title if the setting is empty.

marxjohnson commented on tests/behat/defaultsettings.feature:
```diff
@@ -0,0 +1,26 @@
+    Scenario: Assign some site-wide defaults to the block.
+      Given the following config values are set as admin:
+        | title          | Course Toolkit | block_multiblock |
+        | presentation   | 0              | block_multiblock |
```
It would be good if this test can be extended.  It would be good to see each setting set to something other than the default, and a test that each setting has the desired effect. At the moment its just testing that the default title is displayed, could you add test that the layout and default blocks work too?

james-pearce-bath-ac replied:
I've updated the tests to remove the title test and added tests for the presentation and default blocks.

marxjohnson commented on version.php:
```diff
@@ -24,7 +24,7 @@
 defined('MOODLE_INTERNAL') || die();
 
-$plugin->version   = 2020060104;
+$plugin->version   = 2021111600;
```
Just to avoid confusion, could you bump this to the current date?

james-pearce-bath-ac replied:
I've updated the plugin version with the current date "2022062100"

marxjohnson reviewed (changes requested):
Hi James,
Thanks very much for this, I have added some review comments for you to look at. Its mostly just some things that it would be good to neaten up, plus a couple of changes to keep the default behaviour the same for other users as far as possible.

james-pearce-bath-ac committed:
Pull request fixes

- incremented version number to ‘2022062000’
- fixed language strings
- fixed default values in settings
- fixed checks against config values

james-pearce-bath-ac commented:
I've made some changes to the PR to fix the requests. Please review and let me know if there is anything else that needs to be done.

marxjohnson commented on tests/behat/defaultsettings.feature:
```diff
@@ -15,12 +15,19 @@ Background:
       And I am on "Course 1" course homepage with editing mode on
       And I add the "Multiblock" block
-      Then I should see "Course Toolkit"
+      Then I should see an element with css selector "multiblock-accordion"
```
I don't think this behat step exists? You probably want something like:
`Then ".multiblock.multiblock-accordian" "css_element" should exist`

I dont think it actually creates the accordian element if the block has no content, so you'll need to add a subblock as well in this scenario.

marxjohnson commented on tests/behat/defaultsettings.feature:
```diff
@@ -15,12 +15,19 @@ Background:
+      When I log in as "teacher1"
+      And I am on "Course 1" course homepage with editing mode on
+      And I add the "Multiblock" block
+      Then I should see an element with css selector "multiblock" contains "Calendar"
```
Again, this step doesn't appear to exist, you could do something like:
`Then "Calendar" "text" should exist in the "Multiblock" "block"`

marxjohnson commented on block_multiblock.php:
```diff
@@ -184,10 +183,11 @@ public function get_content() {
         foreach ($presentations as $presentationid => $presentation) {
             array_push($multiblockpresentationoptions, $presentationid);
         }
+        $configuredpresentation = get_config('block_multiblock')->presentation;
```
I would change this to `get_config('block_multiblock', 'presentation')`, otherwise if `presenation` isn't set this might generate an error.

marxjohnson commented on block_multiblock.php:
```diff
@@ -336,15 +336,13 @@ public static function get_valid_presentations(): array {
     public static function get_default_presentation(): string {
         $presentations = static::get_valid_presentations();
         $multiblockpresentationoptions = [];
+        $configuredpresentation = get_config('block_multiblock')->presentation;
```
Same as above.

marxjohnson reviewed (changes requested):
Hi James,
Just a couple of small tweaks to the changes you've made. Also please see my comment regarding the title setting, I didn't mean for you to remove it entirely, sorry! I thought GitHub would highlight the specific line I was talking about, so I wasn't specific enough in my comment.

james-pearce-bath-ac commented:
I've made additional updates to the PR to bring back the block title but make it default to empty and sort the tests.

marxjohnson committed:
Code review fixes

* Moved "addefaultblock" to first-level block_multiblock namespace
* Set the fallback presentation back to 'tabbed-list'
* Fixed defaultsettings behat test
* Removed "Course Toolkit" name from other behat tests

Change-Id: I2eb0dc9b28e2fa18bc489e499c08c8704dd7463a

marxjohnson commented:
Hi James,
To avoid holding this up further, I fixed a couple of remaining issues and merged the branch. I squashed Muhammads, yours and my changes into a single commit each to preseve authorship but keep the commit history tidy. You can see that changes I made in my commit message:
https://github.com/catalyst/moodle-block_multiblock/pull/80/commits/f663501580805da23c0b2ad0805c5a7ef084d9eb

marxjohnson commented:
Thanks very much for your contribution!
